feat: add "offline" SQL backend

jorainer opened the pull request:
- Add `MsBackendOfflineSql` that connects/disconnects to the database for each function call.
- Supports parallel processine issue #9.

jorainer commented:
Does this include also PR#6 @plantton ?

plantton commented:
> Does this include also PR#6 @plantton ?

Yes. It contains PR#6.

jorainer commented on .github/workflows/check-standard.yaml:
```diff
@@ -36,6 +36,16 @@ jobs:
 
       - uses: r-lib/actions/setup-pandoc@master
 
+      - name: Install netcdf (macOS)
```
Just out of curiosity: is the `sysreqs` package not installing these by default?

lgatto replied:
Isn't this an `mzR` requirement?

plantton replied:
I added this dependencies since servers themselves don't contain this library. 

https://github.com/rformassspectrometry/MsBackendSql/issues/9#issue-632642019

jorainer commented on R/MsBackendSqlDb-functions.R:
```diff
@@ -116,24 +120,151 @@ MsBackendSqlDb <- function() {
+    if (missing(i))
+        return(x)
+    if (length(x@index) > 0) {
+        i <- i2index(i, length(x), rownames(x@spectraData))
```
I'm confused - I thought that this backend does **not** store any data in the `x@spectraData` slot but retrieves also the metadata (i.e. the `x@spectraData`) from the database?

plantton replied:
I just notice this PR also including the modifications from yesterday. There're still on the working progress. I will try to solve all the functions by these two days.

lgatto replied:
This seems to be resolved.

jorainer commented on R/MsBackendSqlDb.R:
```diff
@@ -51,10 +52,14 @@ NULL
 #'   to match between the database and the object.
 #' - `@rows` (`integer`): the indices (primary keys) of the data.
 #' - `@columns` (`character`): the names of the columns stored in the database.
+#' - `@index` (`integer`): the index of the rows, which has been labelled by
```
Is this not redundant with `@rows`? I thought that in `@rows` you store the primary keys of the data and that filtering and subsetting changes these (also eventually re-ordering them)?

jorainer commented on R/MsBackendSqlDb.R:
```diff
@@ -76,12 +81,14 @@ setClass("MsBackendSqlDb",
                    rows = "integer",
                    columns = "character",
-                   query = "DBIResult"),
+                   query = "DBIResult",
```
Please also document this slot.

lgatto replied:
@plantton could you provide a description of that slot here please (as a reply to this comment), including a short description of why/how it is used/needed.

jorainer commented on R/MsBackendSqlDb.R:
```diff
@@ -76,12 +81,14 @@ setClass("MsBackendSqlDb",
                    columns = "character",
-                   query = "DBIResult"),
+                   query = "DBIResult",
+                   index = "list"),
```
According to the documentation this is `integer` - please update/correct.

lgatto replied:
The `index` slot seems to have disapeared.

jorainer commented on R/MsBackendSqlDb.R:
```diff
@@ -100,28 +123,365 @@ setValidity("MsBackendSqlDb", function(object) {
+#' 
+setMethod("$", signature = "MsBackendSqlDb", 
+          function(x, name) {
+    if (length(name) == 1) {
```
Can `$` really be called with a `name` of length > 1? I guess in most cases it will be just of length one, so I'm not sure if you really need the `if` `else` here.

lgatto commented:
Careful, there are currently two PR, #10 (this one) and #6, but #6 has additional commit that are missing from this one. Both will need to be merged, hoping that no new conflicts will arise.

plantton commented:
> Careful, there are currently two PR, #10 (this one) and #6, but #6 has additional commit that are missing from this one. Both will need to be merged, hoping that no new conflicts will arise.

I will review them again.

lgatto commented:
> I will review them again.

Focus on one branch now to minimise further conflicts and I will merge both later.

sgibb commented:
While focusing on one branch would be useful I like to suggest to create smaller increment PRs. Reviewing these large nearly daily changing PRs is hard (and for me nearly impossible to follow/catch up).

plantton commented:
> While focusing on one branch would be useful I like to suggest to create smaller increment PRs. Reviewing these large nearly daily changing PRs is hard (and for me nearly impossible to follow/catch up).

Yes, indeed. Laurent also told me to send small PRs instead of a huge one. It would be really tough for reviewers. I will also send small PRs in future.

lgatto commented:
FYI: It looks like your `chmaster` branch (see PR #6) has been deleted.

plantton commented:
> FYI: It looks like your `chmaster` branch (see PR #6) has been deleted.

I actually didn't create this branch on my own repository. I want to focus on the `master` branch.

jorainer committed:
feat: add "offline" SQL backend

- Add `MsBackendOfflineSql` that connects/disconnects to the database for each
  function call.
- Supports parallel processine issue #9.

jorainer committed:
fix: fix unit test
